cafebabel/django_migrations.py

```diff
@@ -1,43 +1,224 @@
+                'language': normalize_language(group_fields['language'])
+            }
+            if data['language'] == 'xx':
+                # It happens with unused tags(?)
+                continue
+            try:
+                tag = Tag.objects.get_or_create(**data)
+            except mongo_errors.NotUniqueError:
+                # We fallback on the old slug in that particular case.
+                data['slug'] = group_fields['slug']
+                try:
+                    tag = Tag.objects.get_or_create(**data)
+                except mongo_errors.NotUniqueError:
+                    del data['name']
+                    tag = Tag.objects.get(**data)
+            summary = group_fields['about']
+            if summary:
+                tag.modify(summary=summary)
+            tags.append(tag)
+    return tags
+
+
+def aggregate_gallery_body(gallery):
+    body = gallery['fields']['body']
+    for image in gallery['images']:
+        image_fields = image['fields']
+        image_src = normalize_image(image_fields['image'])
+        body += ('<figure>'
+                 f'<img src="{image_src}" alt="{image_fields["title"]}">'
+                 f'<figcaption>{image_fields["body"]}</figcaption>'
+                 '</figure>')
+    return body
+
+
+def create_article(old_article):
+    is_gallery = False
+    fields = old_article['fields']
+    status = normalize_status(fields['status'])
+    creation_date = timestamp_to_datetime(fields['created_at'])
+    # Get rid of drafts older than 3 months.
+    if status == 'draft' and creation_date + timedelta(days=90) < NOW:
+        return
+    # Do not consider videos, a couple will be added manually after.
+    if 'video' in old_article:
+        return
+    try:
+        Article.objects.get(archive__pk=old_article['pk'])
+        return
+    except Article.DoesNotExist:
+        pass
+    if 'article' in old_article:
+        article_fields = old_article['article']['fields']
+    elif 'gallery' in old_article:
+        article_fields = old_article['gallery']['fields']
+        is_gallery = True
+    else:
+        click.echo(f'No content: {old_article["pk"]} (spam?)')
+        return
+    # Do not consider articles without body.
+    if not article_fields['body']:
+        return
+    tags = handle_groups(old_article['groups'])
+    data = {
+        'title': fields['title'],
+        'summary': article_fields['original_header'] or '',
+        'body': article_fields['body'],
+        'language': normalize_language(fields['language']),
+        'creation_date': creation_date,
+        'publication_date': timestamp_to_datetime(fields['publication_date']),
+        'author': normalize_author(fields['created_by']),
+        'image_filename': normalize_image(fields['image']),
+        'status': status,
+        'tags': tags or None,
+        'archive': ArticleArchive(
+            pk=old_article['pk'],
+            url=old_article['url'],
+            relateds=fields['relateds']
```

Comment:
Shouldn't we save the new reference instead?

I was thinking that performance-wise it would be consuming but I think you make 2 rounds anyways.

I suppose if you decide to keep this property in the Archive it makes as much sense to keep the archive _pk_ 🤔

Reply:
I think we have to make a third round to be really complete and given that we don't need those yet I prefer to store row old ids and to add references to the `Article` when we will implement the feature for all articles.